engine/consensus/matching/engine.go

```diff
@@ -924,3 +909,53 @@ func (e *Engine) requestPending() error {
 
 	return nil
 }
+
+// check the receipt's data integrity by checking its result has
+// both final statecommitment and initial statecommitment
+func receiptIntegrityCheck(receipt *flow.ExecutionReceipt) (flow.StateCommitment, flow.StateCommitment, error) {
+	final, ok := receipt.ExecutionResult.FinalStateCommitment()
+	if !ok {
+		return nil, nil, fmt.Errorf("execution receipt without FinalStateCommit: %x", receipt.ID())
+	}
+
+	init, ok := receipt.ExecutionResult.InitialStateCommit()
+	if !ok {
+		return nil, nil, fmt.Errorf("execution receipt without InitialStateCommit: %x", receipt.ID())
+	}
+	return init, final, nil
+}
+
+func validateResult(resultsDB storage.ExecutionResults, header *flow.Header, result *flow.IncorporatedResult) error {
+
+	// Retrieve parent result / skip if parent result still unknown:
+	// Before we store a result into the incorporatedResults mempool, we store it in resultsDB.
+	// I.e. resultsDB contains a superset of all results stored in the mempool. Hence, we only need
+	// to check resultsDB. Any result not in resultsDB cannot be in incorporatedResults mempool.
+	previous, err := resultsDB.ByID(result.Result.PreviousResultID)
+	if errors.Is(err, storage.ErrNotFound) {
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("could not get previous result: %w", err)
+	}
+
+	// check sub-graph:
+	// validating the PreviousResultID field
+	// ExecutionResult_X.PreviousResult.BlockID must equal to Block_X.ParentBlockID
+	// for instance: given the following chain
+	// A <- B <- C (ER_A) <- D
+	// a result ER_C with `ID(ER_A)` as its ER_C.Result.PreviousResultID
+	// would be invalid, because `ER_C.Result.PreviousResultID` must be ID(ER_B)
+	valid := header.ParentID == previous.BlockID
+
+	// When `valid` is false, it could either mean this result is invalid, which should be ignored,
+	// or its parent result is invalid, which we could also ignore this result, because we only care
+	// about the first mismatching result, which is the parent result.
+	// In other words, we could ignore this result in both cases.
+	if !valid {
+		return engine.NewInvalidInputErrorf("invalid execution result, parent result block id: %v, executed block's parent id: %v",
+			previous.BlockID, header.ParentID)
+	}
+	return nil
+}
```

Comment:
As Martin pointed out:
> This is already done by the module.ReceiptValidator.

Removing this code would address some of the technical debt listed in issue https://github.com/dapperlabs/flow-go/issues/5137; specifically point 3: 
> * [ ] `matching.Engine` duplicates logic that is now [`ReceiptValidator`](https://github.com/onflow/flow-go/blob/3369dbd1f1e965fae2556f84d677b17297120a05/module/validation/receipt_validator.go)



```suggestion
```